### ISSUE TITLE
Thread 'SortMerger Reading Thread' terminated due to an exception

### ISSUE THREAD

**Issue by twalthr** [stratosphere/stratosphere#551]
By accident, I grouped by a tuple field with type `Double`. The following exception occurs when trying to `groupBy()` by a Double field and using a `reduceGroup()` afterwards:

<code>java.lang.Exception: The data preparation for task 'Reduce(eu.stratosphere.example.java.relational.TPCHQuery1$1)' , caused an error: Error obtaining the sorted input: Thread 'SortMerger Reading Thread' terminated due to an exception: null
    at eu.stratosphere.pact.runtime.task.RegularPactTask.run(RegularPactTask.java:481)
    at eu.stratosphere.pact.runtime.task.RegularPactTask.invoke(RegularPactTask.java:374)
    at eu.stratosphere.nephele.execution.RuntimeEnvironment.run(RuntimeEnvironment.java:345)
    at java.lang.Thread.run(Thread.java:744)
Caused by: java.lang.RuntimeException: Error obtaining the sorted input: Thread 'SortMerger Reading Thread' terminated due to an exception: null
    at eu.stratosphere.pact.runtime.sort.UnilateralSortMerger.getIterator(UnilateralSortMerger.java:635)
    at eu.stratosphere.pact.runtime.task.RegularPactTask.getInput(RegularPactTask.java:1113)
    at eu.stratosphere.pact.runtime.task.ReduceDriver.prepare(ReduceDriver.java:86)
    at eu.stratosphere.pact.runtime.task.RegularPactTask.run(RegularPactTask.java:476)
    ... 3 more
Caused by: java.io.IOException: Thread 'SortMerger Reading Thread' terminated due to an exception: null
    at eu.stratosphere.pact.runtime.sort.UnilateralSortMerger$ThreadBase.run(UnilateralSortMerger.java:826)
Caused by: java.lang.UnsupportedOperationException
    at eu.stratosphere.api.common.typeutils.base.DoubleComparator.putNormalizedKey(DoubleComparator.java:58)
    at eu.stratosphere.api.common.typeutils.base.DoubleComparator.putNormalizedKey(DoubleComparator.java:1)
    at eu.stratosphere.api.java.typeutils.runtime.TupleSingleFieldComparator.putNormalizedKey(TupleSingleFieldComparator.java:101)
    at eu.stratosphere.api.java.typeutils.runtime.TupleSingleFieldComparator.putNormalizedKey(TupleSingleFieldComparator.java:1)
    at eu.stratosphere.api.java.typeutils.runtime.ReferenceWrappedComparator.putNormalizedKey(ReferenceWrappedComparator.java:93)
    at eu.stratosphere.api.java.typeutils.runtime.ReferenceWrappedComparator.putNormalizedKey(ReferenceWrappedComparator.java:1)
    at eu.stratosphere.pact.runtime.sort.NormalizedKeySorter.write(NormalizedKeySorter.java:269)
    at eu.stratosphere.pact.runtime.sort.UnilateralSortMerger$ReadingThread.go(UnilateralSortMerger.java:1056)
    at eu.stratosphere.pact.runtime.sort.UnilateralSortMerger$ThreadBase.run(UnilateralSortMerger.java:822)</code>

I don't know if grouping over a double key makes much sense but anyway the exception is not user friendly and should be improved.

**Comment by StephanEwen** [comment]
The system thinks it can use normalized keys, but double types do not
implement them. Can you check whether the double comparator returns true on
the method that declares whether normalized keys are enabled?

**Comment by twalthr** [comment]
The `DoubleComparator` returns false. I think the problem lies in `NormalizedKeySorter` because the support is checked in the constructor but in `write(T record)` 

```
this.comparator.putNormalizedKey(record, this.currentSortIndexSegment, this.currentSortIndexOffset + OFFSET_LEN, this.numKeyBytes);
```

is called anyway. Would it be ok to simply remove the thrown `UnsupportedException` in `putNormalizedKey` of the comparator?

**Comment by StephanEwen** [comment]
The exception is there for a reason. Simply doing nothing screws up the
behavior. The issue is to figure out why the system tries to use normalized
keys in the first place. Can you debug that?

**Comment by fhueske** [comment]
We had a look at this.
The problem is that the NormalizedKeySorter always calls the putNormalizedKey() method, even if normalized keys are not supported. 
In this case, the NormalizedKeySorter sets the key-length to 0 and expects that the function call doesn't do anything. However, the DoubleComparator throws an exception and causes the failure.

**Comment by StephanEwen** [comment]
Okay, so a quick fix is to do nothing on 0 length calls. We should fix the
behavior of the sorter, though, that it does not even try to put normalized
keys in the first place.

**Comment by fhueske** [comment]
Already done, see #560 :smile:

**Comment by fhueske** [comment]
Fixed in #560
